src/box/field_default_func.c

```diff
@@ -0,0 +1,134 @@
+ */
+#include "port.h"
+#include "engine.h"
+#include "tt_static.h"
+#include "tuple_format.h"
+#include "field_default_func.h"
+
+/**
+ * Find `default_func` in func cache.
+ * Return NULL if something went wrong (diag is set).
+ */
+struct func *
+field_default_func_find(struct field_default_func *default_func)
+{
+	struct func *func = func_by_id(default_func->id);
+	if (func == NULL) {
+		diag_set(ClientError, ER_NO_SUCH_FUNCTION,
+			 int2str(default_func->id));
+	}
+	return func;
+}
+
+/**
+ * Check `func` for compliance with the field default function rules.
+ *
+ * Returns 0 on success. On error, sets diag and returns -1.
+ * `fieldno` is used for better error message.
+ */
+static int
+field_default_func_verify(struct func *func, uint32_t fieldno)
+{
+	const char *func_name = func->def->name;
+
+	if (func->def->language == FUNC_LANGUAGE_LUA &&
+	    func->def->body == NULL) {
+		diag_set(ClientError, ER_WRONG_SPACE_FORMAT,
+			 fieldno + TUPLE_INDEX_BASE,
+			 tt_sprintf("field default Lua function '%s' must have "
+				    "persistent body", func_name));
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * Function that calls field default function `default_func`.
+ * `arg_data` points to msgpack data with the function argument.
+ * `ret_data` is pointed to msgpack data with the function return value,
+ * allocated on current fiber's region.
+ *
+ * Returns 0 on success. On error, sets diag and returns -1.
+ */
+static int
+field_default_func_call(struct field_default_func *default_func,
+			const char *arg_data, uint32_t arg_size,
+			const char **ret_data, uint32_t *ret_size)
+{
+	struct port out_port, in_port;
+	port_c_create(&in_port);
+	if (arg_data != NULL)
+		port_c_add_mp(&in_port, arg_data, arg_data + arg_size);
+
+	int rc = func_call_no_access_check(default_func->holder.func,
+					   &in_port, &out_port);
+	port_destroy(&in_port);
+	if (rc != 0) {
+		diag_log();
+		diag_clear(diag_get());
+		diag_set(ClientError, ER_FIELD_DEFAULT_FUNC,
+			 default_func->holder.func->def->name,
+			 "failed");
+		return -1;
+	}
+
+	*ret_data = port_get_msgpack(&out_port, ret_size);
+	assert(mp_typeof(**ret_data) == MP_ARRAY);
+	unsigned ret_count = mp_decode_array(ret_data);
+	if (ret_count != 1) {
+		diag_set(ClientError, ER_FIELD_DEFAULT_FUNC,
+			 default_func->holder.func->def->name,
+			 tt_sprintf("returned %u values, expected 1",
+				    ret_count));
+		return -1;
+	}
+	*ret_size -= mp_sizeof_array(ret_count);
+	port_destroy(&out_port);
+	return 0;
+}
+
+void
+field_default_func_unpin(struct field_default_func *default_func)
+{
+	func_unpin(&default_func->holder);
+}
+
+void
+field_default_func_pin(struct field_default_func *default_func)
+{
+	struct func *func = field_default_func_find(default_func);
+	func_pin(func, &default_func->holder, FUNC_HOLDER_FIELD_DEFAULT);
+}
+
+void
+field_default_func_destroy(struct field_default_func *default_func)
+{
+	if (default_func->holder.func != NULL)
+		field_default_func_unpin(default_func);
```

Comment:
This line isn't covered.

Reply:
I'm not sure if `field_default_func_destroy()` is needed in `tuple_field_delete()` or not.

There is a similar code in `tuple_field_delete()` for **field** constraints, maybe it is redundant.
```c
	for (uint32_t i = 0; i < field->constraint_count; i++)
		field->constraint[i].destroy(&field->constraint[i]);
```

On the other hand, `tuple_format_destroy()` just frees **tuple** constraints without calling destructors.

Reply:
Removed this code, added `assert`s instead.